src/openms/source/CHEMISTRY/Tagger.cpp

```diff
@@ -0,0 +1,201 @@
+     // This assumes both fragments had the same charge,
+     // but ignores whether ions in the longer sequence all have the same charge.
+     // This could lead to inconsistencies where an ion is considered to have different charges when compared to its left and right neighbors in the sequence (TODO?).
+     // This looks at the difference between two charged ions, so the proton mass is the same on both ions and can be ignored here.
+     for (size_t charge = min_charge_; charge <= max_charge_; ++charge)
+     {
+       double charged_m = m / charge;
+       if (fabs(left->first - charged_m) < delta) return left->second;
+     }
+     return " ";
+  }
+
+  void Tagger::getTag_(StringList & tag, const std::vector<float>& mzs, const size_t i, std::set<std::string>& tags) const
+  {
+    const size_t N = mzs.size();
+    size_t j = i + 1;
+    // recurse for all peaks in distance < max_gap
+    while (j < N)
+    {
+      if (tag.size() == max_tag_length_) { return; } // maximum tag size reached? - continue with next parent
+
+      const float gap = mzs[j] - mzs[i];
+      //cout << i << "\t" << j << "\t" << mzs[i] << "\t" << mzs[j] << "\t" << gap << endl;
+      if (gap > max_gap_) { return; } // already too far away - continue with next parent
+      const std::string aa = getAAByMass_(gap);
+      if (aa == " ") { ++j; continue; } // can't extend tag
+      tag.push_back(aa);
+      getTag_(tag, mzs, j, tags);
+      // each tag is build as a StringList, so that push_back(), pop_back() and size() work as intended even with modified residues, e.g. "M(Oxidation)"
+      // and then the tag is reported as a string by collapsing it here using concatenate()
+      if (tag.size() >= min_tag_length_) tags.insert(ListUtils::concatenate(tag));
+
+      // if aa is "L", then also add "I" as an alternative residue and extend the tag again
+      // this will add redundancy, (and redundant runtime) but we avoid dealing with J and ambigous matching to I and L later on
+      // TODO would it be more efficient to only generate tags with L and after they are finished make copies with L exchanged for I?
+      // TODO Then we would need to generate all possible combinations, if multiple L are in the tag
+      if (aa == "L")
+      {
+        tag.pop_back();
+        tag.push_back("I");
+        getTag_(tag, mzs, j, tags);
+      }
+      if (tag.size() >= min_tag_length_) tags.insert(ListUtils::concatenate(tag));
+      tag.pop_back();  // remove last string
+      ++j;
+    }
+  }
+
+  Tagger::Tagger(size_t min_tag_length, float ppm, size_t max_tag_length, size_t min_charge, size_t max_charge, StringList fixedModNames, StringList varModNames)
+  {
+    ppm_ = ppm;
+    min_tag_length_ = min_tag_length;
+    max_tag_length_ = max_tag_length;
+    min_charge_ = min_charge;
+    max_charge_ = max_charge;
+
+    const std::set<const Residue*> aas = ResidueDB::getInstance()->getResidues();
```

Comment:
probably better to choose a residue set here (check the parameters of getResidues) that is not ambigious